### PR TITLE
Implement TypeScript support (closes #408)

### LINKED ISSUE #408: TypeScript compiler
Make TypeScript first class citizen by automatically compiling `.ts` files like we do with ES6. This can be  done (and more likely should) as standalone plugin.

### PR DESCRIPTION
**Pull request by inikulin**
This enables built-in compilation of `.ts` files and type definitions for TestCafe. You need to `import 'testcafe';` (in case if you don't import e.g. `Selector` already) to enable typings, editor support and make test files compilable:
```ts
import `testcafe`;

fixture('My fixture');

test('My test', async t => {
});
```

Note that tagged template literal syntax is not available for `fixture` and `page`. Use function syntax instead.

To try code completion and API hints in VSCode run `npm install @inikulin/testcafe` and then `import '@inikulin/testcafe'` in your test file.

\cc @AlexanderMoskovkin @helen-dikareva @AndreyBelym 

Note that we've dropped Node 0.10 support with this feature.

### CODE REVIEW

File: src/compiler/test-file/api-based.js [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    cleanUp () {
+        this.cache = {};
+    }
+}
+
+
+// Const
+APIBasedTestFileCompilerBase.EXPORTABLE_LIB_PATH = join(__dirname, '../../api/exportable-lib');
```
Comment (helen-dikareva) [inline src/compiler/test-file/api-based.js]: we can use `static get` for this

File: src/compiler/test-file/formats/typescript.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            target:          'ES6',
+            lib:             ['lib.es6.d.ts'],
+            baseUrl:         __dirname,
+            paths:           { testcafe: ['../../../../ts-defs/index.d.ts'] }
+        };
+    }
+
+    static _reportErrors (diagnostics) {
+        // NOTE: lazy load the compiler
+        var ts     = require('typescript');
```
Comment (helen-dikareva) [inline src/compiler/test-file/formats/typescript.js]: may be use `get` for `ts`

File: package.json [hunk 3]
```diff
@@ -92,17 +97,19 @@
...
-    "testcafe-legacy-api": "2.0.2",
+    "testcafe-legacy-api": "3.0.0",
     "testcafe-reporter-json": "^2.1.0",
     "testcafe-reporter-list": "^2.1.0",
     "testcafe-reporter-minimal": "^2.1.0",
     "testcafe-reporter-spec": "^2.1.1",
     "testcafe-reporter-xunit": "^2.1.0",
     "time-limit-promise": "^1.0.2",
     "tree-kill": "^1.1.0",
+    "typescript": "^2.2.2",
```
Comment (RichiCoder1) [inline package.json]: Is it possible to make typescript a peer dependency? Likely many projects are already going to have typescript pulled in as a dependency.

Comment (inikulin) [inline package.json]: @RichiCoder1  Is there any problem with having ts as dependency? Unlike e.g. babel it has zero dependencies and package size is quite small. Moreover, if you already have ts installed package manager will more likely link already installed version.

File: src/compiler/test-file/api-based.js (hunk 1 above)
Comment (helen-dikareva) [inline src/compiler/test-file/api-based.js]: I think declare a const and use `static get` will be more in es6 style

File: src/compiler/index.js [hunk 4]
```diff
@@ -1,32 +1,48 @@
...
-            new LegacyCompiler(hammerhead.processScript),
-            this.esNextCompiler,
-            new RawFileCompiler()
-        ];
+        this.sources = sources;
+
+        Compiler._setupSourceMapsSupport();
+    }
+
+    static getSupportedTestFileExtensions () {
```
Comment (helen-dikareva) [inline src/compiler/index.js]: ok, do not mind

File: src/compiler/test-file/formats/typescript.js (hunk 2 above)
Comment (helen-dikareva) [inline src/compiler/test-file/formats/typescript.js]: forget it)

File: package.json (hunk 3 above)
Comment (RichiCoder1) [inline package.json]: True, but it does open yourself up to issues if Typescript releases a major version, and the typescript package is yet another package for end users. That being said, I don't feel strongly about this :)

Comment (inikulin) [inline package.json]: Yeah, but we'll be open for such issues anyway, i.e. programmatical API that we use could change with a major version bump.

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by AlexanderMoskovkin** [comment]
The only functional test is failed. It relates to the `pressKey` automation in Safari. It failed in all current builds (pull requests and master branch), so it's not related to the current pull request, but I think it's because browser version was updated on BrowserStack. We are investigating it.
So, I think we can ignore this test and merge the pull request if there are no changes requests here.

**Pull request merged**